### PR TITLE
fix(prow/apis/prowjobs,prow/crier): fix job state udpate for jenkins job

### PR DESCRIPTION
**Pull request by wuhuizuo**
## Why
When enabled the reporting for Jenkins-operator, there will be two comments when the job failed.
Crier component have a issue for updating report for Jenkins jobs:
- The Prow job state will be pending when Jenkins enqueued the build. At this state, there are none detail url in report.
- The state will be updated only when the build failed or succeed or aborted. User can not get the detail url when running.
image

## What I changed

- Extract to a method to judge whether need to update the report.
- Add a field to record previous state description for `jenkins` agent.
- Report when build state description changed for `jenkins` agent.

### CODE REVIEW

**Review by jmguzik — COMMENTED** [review]
This change is touching prow core components. Even if the change is affecting only Jenkins feature, some tests are recommended.

**Comment by wuhuizuo** [comment]
> This change is touching prow core components. Even if the change is affecting only Jenkins feature, some tests are recommended.

Thanks!
I will add some tests for it.

**Comment by wuhuizuo** [comment]
> This change is touching prow core components. Even if the change is affecting only Jenkins feature, some tests are recommended.

@jmguzik PTAL

File: prow/apis/prowjobs/v1/types.go [hunk 1]
```diff
@@ -1018,6 +1019,23 @@ func (j *ProwJob) ClusterAlias() string {
...
 
+// NeedReportState return true if it need to report the new state for the job.
+func (j *ProwJob) NeedReportState(reporter string) bool {
+	if j.Status.PrevReportStates[reporter] != j.Status.State {
+		return true
+	}
+
+	if j.Spec.Agent != JenkinsAgent {
+		return false
+	}
```
Comment (petr-muller) [inline prow/apis/prowjobs/v1/types.go:1030]: Is there a good reason why this behavior should be limited to Jenkins controller? If prow thinks the status should have a different description on the same state, then it should propagate no matter if the job is jenkins or anything else, no?

**Review by petr-muller — COMMENTED** [review]
If the motivation for this is that the URL is not propagated to GitHub `details` link, then I'd expect that to be handled somehow. I guess the fix improves the situation by assuming that if the URL appears then description will likely change too, but what if it doesnt?

File: prow/apis/prowjobs/v1/types.go (hunk 1 above)
Comment (wuhuizuo) [reply prow/apis/prowjobs/v1/types.go:1030]: If you agree, I cloud add it , but current only Jenkins agent has this case: 
```text
Prow job: triggered ------> pending ---------------------------------------------------> succeed or failed
                                                 \                                                                                                                       /
                                                  \                                                                                                                     /
                            Jenkins job: queued(no job url) --> running(job description and url changed) --> succeed or failed(only description changed)
```

Comment (wuhuizuo) [reply prow/apis/prowjobs/v1/types.go:1030]: Perhaps a new state will make it simpler: "queuing", but other agents do not necessarily have the same state sets.

Comment (petr-muller) [reply prow/apis/prowjobs/v1/types.go:1030]: Fair enough, I guess it's fine. I'm still curious about the main review comment.

**Comment by wuhuizuo** [comment]
> If the motivation for this is that the URL is not propagated to GitHub `details` link, then I'd expect that to be handled somehow. I guess the fix improves the situation by assuming that if the URL appears then description will likely change too, but what if it doesnt?

@petr-muller 
Jenkins build's url will be appeared when the state are transformed from `queued` to `running`, it will not be changed  after that.

For other agent like `kubernetes` and `pipeline` have not such situation, so I think the change should be limited to `jenkins` agent and should not spread to other.

**Review by petr-muller — APPROVED** [review]

**Comment by petr-muller** [comment]
right, this updates the CRD so it needs an owner from oncall

/assign @chases2

**Comment by wuhuizuo** [comment]
ping @chases2

**Comment by wuhuizuo** [comment]
@chaodaiG Looking forward to your reviewing!

**Review by listx — COMMENTED** [review]
Please remove the [merge commit in your PR](https://github.com/kubernetes/test-infra/pull/28783/commits/3d6fe51b9e67abe2e4f7e1672e2e33318be8bc8b). PRs shouldn't have merge commits in them because they make the history messy.

Also I will be going on PTO for 3 weeks starting 5/15, so expect delays in reviewing this.

**Comment by wuhuizuo** [comment]
Or can we introduce a new job state to make it simpler?

**Pull request closed without merging**